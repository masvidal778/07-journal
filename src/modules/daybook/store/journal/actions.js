//Accions asincrones que poden cridar a una mutaciÃ³
import journalApi from "@/api/journalApi";

export const loadEntries = async ( { commit } ) => {
    const { data } = await journalApi.get('/entries.json')

    const entries = []
    for( let id of Object.keys( data ) ){
        entries.push({
            id,
            ...data[id]
        })
    }

    commit('setEntries', entries)

}


export const updateEntries = async (/*{ commit }*/) => {

}

export const createEntries = async (/*{ commit }*/) => {

}
