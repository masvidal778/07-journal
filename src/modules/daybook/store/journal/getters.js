//Portar informaciÃ³ de l'state que es pot processar
//export const myGetter = ( state ) => {
//return state
//}

export const getEntriesByTerm = ( /*state*/ ) => {

}

export const getEntriesById = ( /*state*/ ) => {

}