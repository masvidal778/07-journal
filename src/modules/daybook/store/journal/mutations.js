//Síncrones. Modifiquen l'state
export const setEntries =  ( state, entries ) => {

    state.entries = [ ...state.entries, ...entries ]
    state.isLoading = false

}

export const updateEntry =  ( state, entryUpdated ) => { //receive entry updated

    //crea un nou array on només hi ha els id de l'store i els compara amb els id de l'entryUpdated
    const index = state.entries.map( e => e.id ).indexOf( entryUpdated.id )
    state.entries[ index ] = entryUpdated


}

export const addEntry =  ( /*state*/ ) => {

}