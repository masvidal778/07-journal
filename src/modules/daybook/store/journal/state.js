//Ã‰s reactiu. Quan canvia es notifica a tots els components que
//observen quan es canvia l'state'

export default () => ({
    isLoading: true,
    entries: [
        /*
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Aliquam tristique erat massa, at commodo mauris vulputate vel. Aenean ut tincidunt magna, quis bibendum risus. Etiam dignissim, ipsum vitae accumsan vestibulum, ante velit consectetur est, quis scelerisque enim libero vel massa',
            picture: null
        }

         */
    ]
})