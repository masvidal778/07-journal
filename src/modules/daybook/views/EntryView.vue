<script>
import {defineAsyncComponent, defineComponent} from 'vue'

export default defineComponent({
  name: "EntryView",
  components: {
    FabNew: defineAsyncComponent( () => import('../components/FabNew.vue') )
  }
})
</script>

<template>
<div class="entry-title d-flex justify-content-between p-2">
  <div>
    <span class="text-success fs-3 fw-bold">15</span>
    <span class="mx-1 fs-3">Setembre</span>
    <span class="mx-2 fs-4 fw-light">2023, divendres</span>
  </div>

  <div>
    <button class="btn btn-danger mx-2">
      Esborrar
      <i class="fa fa-trash-all"></i>
    </button>

    <button class="btn btn-primary">
      Pujar fotografia
      <i class="fa fa-upload"></i>
    </button>
  </div>
</div>

  <hr>

  <div class="d-flex flex-column px-3 h-75">
    <textarea
        name=""
        id=""
        cols="30"
        rows="10"
        placeholder="QuÃ¨ ha passat avui?"
    ></textarea>
  </div>

  <FabNew/>

  <img
      src="https://static01.nyt.com/images/2022/11/29/science/00tb-cats1/00tb-cats1-mediumSquareAt3X.jpg"
      alt="entry-picture"
      class="img-thumbnail"
  >
</template>

<style lang="scss" scoped>
textarea{
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus{
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>