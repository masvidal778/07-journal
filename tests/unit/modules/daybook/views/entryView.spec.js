import {shallowMount} from "@vue/test-utils";
import { createStore } from "vuex";

import Swal from 'sweetalert2'

import journal from "@/modules/daybook/store/journal";
import EntryView from "@/modules/daybook/views/EntryView.vue";
import { journalState } from "../../../mock-data/test-journal-state";

const createVuexStore = ( initialState ) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...initialState }
            }
        }
    })

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Testing EntryView', () => {

    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: 'ABC123'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        } )
    })

    test("should to disconnect user because id doesn't exist", () => {

        const wrapper = shallowMount( EntryView, {
            props: {
                id: "This ID doesn't exist on the store"
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        } )

        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })

    })

    test("entry should be showed correctly", () => {

        expect( wrapper.html() ).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()

    })

    test('should erase entry', (done) => {

        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))

        wrapper.find('.btn-danger').trigger('click')

        expect(Swal.fire).toHaveBeenCalledWith({
            title: "Està segur que vol esborrar l'entrada?",
            text: "Un cop esborrada, no es podrà recuperar",
            showDenyButton: true,
            confirmButtonText: "Sí, vull esborrar l'entrada"
        })

        setTimeout( () => {
            expect( store.dispatch).toHaveBeenCalledWith("journal/deleteEntries", "ABC123")
            expect( mockRouter.push ).toHaveBeenCalled()
            done()
        }, 1)

    })

})