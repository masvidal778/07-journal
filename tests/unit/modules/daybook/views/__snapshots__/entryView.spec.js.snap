// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing EntryView entry should be showed correctly 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">18</span><span class="mx-1 fs-3">Juliol</span><span class="mx-2 fs-4 fw-light">2021, Diumenge</span></div><input type="file" accept="image/png, image/jpg" style="display: none;">
  <div><button class="btn btn-danger mx-2"> Esborrar <i class="fa fa-trash-all"></i></button><button class="btn btn-primary"> Pujar fotografia <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="QuÃ¨ ha passat avui?"></textarea></div>
<img src="https://res.cloudinary.com/dqkxcuput/image/upload/v1697048344/wibhrg2otqmhoqdcg3kc.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-new-stub icon="fa-save"></fab-new-stub>
`;
